### PR TITLE
PMM-5492 Adding pprof query parameter to /logs.zip

### PR DESCRIPTION
**Pull request by pkadej**
[PMM-5492](https://jira.percona.com/browse/PMM-5492) Add Pprof data to logs.zip

Build: [SUBMODULES-2574](https://github.com/Percona-Lab/pmm-submodules/pull/2574)

### CODE REVIEW

**Review by BupycHuk — DISMISSED** [review]

File: api/serverpb/json/header.json [hunk 1]
```diff
@@ -17,6 +17,15 @@
         "description": "Returns the PMM Server logs.",
         "summary": "Logs",
         "operationId": "Logs",
+        "parameters": [
+          {
+            "type": "integer",
```
Comment (BupycHuk) [inline api/serverpb/json/header.json:22]: can we replace it with a `boolean` type?

**Review by BupycHuk — CHANGES_REQUESTED** [review]

File: api/serverpb/json/header.json (hunk 1 above)
Comment (pkadej) [reply api/serverpb/json/header.json:22]: @BupycHuk 
In JIRA ticket, we have: "User calls /logs.zip endpoint with `pprof=1` query parameter" that's why I used integer here. Shall I update JIRA ticket as well?

Comment (BupycHuk) [reply api/serverpb/json/header.json:22]: boolean supports `pprof=1` as well and makes more sense

Comment (pkadej) [reply api/serverpb/json/header.json]: Thank you for clarification. Changed.

**Comment by pkadej** [comment]
Updated with changes from pmm-managed's PR (which has archived repository): https://github.com/percona/pmm-managed/pull/1117/commits

File: admin/commands/summary.go [hunk 2]
```diff
@@ -153,9 +153,9 @@ func addClientData(ctx context.Context, zipW *zip.Writer) {
 }
 
 // addServerData adds logs.zip from PMM Server to zip file.
-func addServerData(ctx context.Context, zipW *zip.Writer) {
+func addServerData(ctx context.Context, zipW *zip.Writer, usePprof bool) {
 	var buf bytes.Buffer
-	_, err := client.Default.Server.Logs(&server.LogsParams{Context: ctx}, &buf)
+	_, err := client.Default.Server.Logs(&server.LogsParams{Context: ctx, Pprof: &usePprof, HTTPClient: nil}, &buf)
```
Comment (artemgavrilov) [inline admin/commands/summary.go:158]: We can omit `HTTPClient: nil`

File: api/serverpb/json/client/server/logs_parameters.go [hunk 3]
```diff
@@ -58,6 +59,12 @@ func NewLogsParamsWithHTTPClient(client *http.Client) *LogsParams {
    Typically these are written to a http.Request.
 */
 type LogsParams struct {
+	/* Pprof.
+
+	   Include performance profiling data,
+	*/
```
Comment (artemgavrilov) [inline api/serverpb/json/client/server/logs_parameters.go:65]: Can we use one-line comment here?

File: api/serverpb/json/client/server/logs_parameters.go [hunk 4]
```diff
@@ -58,6 +59,12 @@ func NewLogsParamsWithHTTPClient(client *http.Client) *LogsParams {
...
+	Pprof *bool
```
Comment (artemgavrilov) [inline api/serverpb/json/client/server/logs_parameters.go:66]: What is the reason to have it as a pointer?

File: api/serverpb/json/client/server/logs_parameters.go [hunk 5]
```diff
@@ -111,13 +118,40 @@ func (o *LogsParams) SetHTTPClient(client *http.Client) {
...
 	var res []error
 
+	if o.Pprof != nil {
+
+		// query param pprof
+		var qrPprof bool
+
+		if o.Pprof != nil {
+			qrPprof = *o.Pprof
+		}
```
Comment (artemgavrilov) [inline api/serverpb/json/client/server/logs_parameters.go:146]: `o.Pprof != nil` checked twice

File: api/serverpb/json/client/server/logs_parameters.go [hunk 6]
```diff
@@ -111,13 +118,40 @@ func (o *LogsParams) SetHTTPClient(client *http.Client) {
...
+		qPprof := swag.FormatBool(qrPprof)
+		if qPprof != "" {
```
Comment (artemgavrilov) [inline api/serverpb/json/client/server/logs_parameters.go:148]: `swag.FormatBool` can return only `true` or `false`. No need to check for emptiness

File: managed/utils/pprof/pprof.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+	"fmt"
+	"runtime"
+	"runtime/pprof"
+	"runtime/trace"
+	"time"
+)
+
+// Profile responds with the pprof-formatted cpu profile.
+// Profiling lasts for duration specified in seconds.
+func Profile(duration time.Duration) ([]byte, error) {
```
Comment (artemgavrilov) [inline managed/utils/pprof/pprof.go:30]: I think we should have context in arguments and terminate this function execution if context is cancelled. 


Same for `Trace`

File: managed/utils/pprof/pprof_config.go [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+package pprof
+
+import "time"
+
+// Config pprof settings.
+type Config struct {
+	ProfileDuration time.Duration `yaml:"profile_duration"` //nolint:tagliatelle
+	TraceDuration   time.Duration `yaml:"trace_duration"`   //nolint:tagliatelle
```
Comment (artemgavrilov) [inline managed/utils/pprof/pprof_config.go:24]: I guess we can remove yaml tags here

File: managed/utils/pprof/pprof_config.go [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+// Init pprof config init.
+func (c *Config) Init() {
+}
```
Comment (artemgavrilov) [inline managed/utils/pprof/pprof_config.go:29]: We can remove it.

File: managed/utils/pprof/pprof_config.go [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+package pprof
+
+import "time"
+
+// Config pprof settings.
+type Config struct {
```
Comment (artemgavrilov) [inline managed/utils/pprof/pprof_config.go:22]: This config defined in `utils/pprof` but used only in `services/supervisord`. Let's move it there.

**Review by artemgavrilov — CHANGES_REQUESTED** [review]

File: api/serverpb/json/client/server/logs_parameters.go (hunk 3 above)
Comment (pkadej) [reply api/serverpb/json/client/server/logs_parameters.go:65]: Well, this file is generated automaticly...

File: api/serverpb/json/client/server/logs_parameters.go (hunk 5 above)
Comment (pkadej) [reply api/serverpb/json/client/server/logs_parameters.go:146]: File generated automaticly.

File: api/serverpb/json/client/server/logs_parameters.go (hunk 3 above)
Comment (artemgavrilov) [reply api/serverpb/json/client/server/logs_parameters.go:65]: Damn 😖

File: admin/commands/summary.go (hunk 2 above)
Comment (pkadej) [reply admin/commands/summary.go:158]: Yes we can; i have added it because of linter warning.

Comment (artemgavrilov) [reply admin/commands/summary.go:158]: What linter? Before we disabled `exhaustivestruct` liter exact for such useless warnings.

File: managed/utils/pprof/pprof.go (hunk 7 above)
Comment (pkadej) [reply managed/utils/pprof/pprof.go]: Done.

File: managed/utils/pprof/pprof_config.go (hunk 8 above)
Comment (pkadej) [reply managed/utils/pprof/pprof_config.go]: Removed.

File: managed/utils/pprof/pprof_config.go (hunk 9 above)
Comment (pkadej) [reply managed/utils/pprof/pprof_config.go]: Removed.

File: managed/utils/pprof/pprof_config.go (hunk 10 above)
Comment (pkadej) [reply managed/utils/pprof/pprof_config.go]: Moved.

**Review by artemgavrilov — APPROVED** [review]

**Comment by michal-kralik** [comment]
@pkadej can you please take a look at why the managed tests are failing and increase the timeout if needed?

**Review by michal-kralik — APPROVED** [review]

**Pull request merged**